src/instructlab/training/utils.py

```diff
@@ -781,3 +782,103 @@ def set_random_seed(seed):
+
+def save_checkpoint(args, accelerator: Accelerator, model, tokenizer, samples_seen, is_lora: bool, epoch: int=None, hf_format: bool=True, full_state: bool=False, )->None:
+    if hf_format:
+        save_hf_format_accelerate(
+            args=args,
+            model=model,
+            accelerator=accelerator,
+            tokenizer=tokenizer,
+            samples_seen=samples_seen,
+            is_lora=is_lora
+        )
+
+    if full_state:
+        save_full_state(
+            args=args,
+            accelerator=accelerator,
+            tokenizer=tokenizer,
+            samples_seen=samples_seen,
+            is_lora=is_lora,
+            epoch=epoch 
+        )
+
+
+def save_full_state(args, accelerator, tokenizer, samples_seen, is_lora, epoch):
+    """
+    Saves model, optimizer, and lr_scheduler state.
+    TODO: save model config
+    TODO: save tokenizer
+    TODO: handle LoRA
+    TODO: handle granite
+    """
+    if is_lora:
+        raise NotImplementedError("Can't save full state for LoRA at the moment.")
+
+    if args.is_granite:
+        raise NotImplementedError("Can't save full state for Granite models yet.")
+
+    output_dir = Path(args.output_dir) / "full_state" / f"samples_{samples_seen}"
+    log_rank_0(f"\033[93mSaving full model state in {output_dir}\033[0m", to_print=True)
+    
+    # patch FSDP state dict method so it works correctly.
+    def _get_state_dict_patched(model, unwrap=False):
+        return get_state_dict_unpatched(model, unwrap=unwrap)
+    if args.distributed_training_framework == "fsdp":
+        get_state_dict_unpatched = accelerator.get_state_dict
+        accelerator.get_state_dict = _get_state_dict_patched
+    
+    accelerator.save_state(
+        output_dir=output_dir,
+        # max_shard_size="5GB",
+        # safe_serialization=True,
+    )
+
+    # save metadata file for current training status
+    if accelerator.is_main_process:
+        # TODO: should we set the global_step here rather than calculating global_step
+        #   based on samples_seen?
+        metadata = {"current_epoch": epoch, "samples_seen": samples_seen}
+        torch.save(metadata, output_dir / "training_metadata.json")
+        log_rank_0(f"\033[93mSaving training state: {metadata}\033[0m", to_print=True)
+
+    log_rank_0(f"\033[93mModel state saved in: {output_dir}\033[0m", to_print=True)
+
+    # cleanup
+    if args.distributed_training_framework == "fsdp":
+        accelerator.get_state_dict = get_state_dict_unpatched 
+
+def load_latest_full_state(args, accelerator)->None:
+    """
+    Loads accelerator state from most recently saved checkpoint
+    in `output_dir/full_state`. 
+    """
+    output_dir = Path(args.output_dir) / "full_state"
+
+    # dist.breakpoint()
+    if not output_dir.is_dir():
+        return
+
+    # picks checkpoint with the largest number of samples seen, by name.
+    checkpoint_list = sorted(list(output_dir.iterdir()), reverse=True)
```

Comment:
Is this necessarily true? If the list is a list of strings then how does it handle a case like `samples_1000` vs `samples_99999`

Reply:
Usually hash of the string or sum of the string characters, so 9999 > 1000.

Reply:
I would advise passing a custom comparison function to sorted here, see the following code snippet:

```py
lst = ["hf_format_1", "hf_format_10", "hf_format_2"]
sorted_lst = sorted(lst)
print(sorted_lst)
> ['hf_format_1', 'hf_format_10', 'hf_format_2']
```

You can pass a custom function that simply splits on the `_` character and casts the last part of the string into a number, this should do what you want it to.

Reply:
This will get `sorted` to work how you want it to:
```suggestion
    checkpoint_list = sorted(list(output_dir.iterdir()), reverse=True, key=lambda d: int(d.split('_')[-1]))
```